src/modules/hyprland/workspaces.cpp

```diff
@@ -551,136 +585,75 @@ void Workspaces::removeWorkspace(std::string const &name) {
-            if (monitor.isString() && monitor.asString() == m_bar.output->name) {
-              m_persistentWorkspacesToCreate.emplace_back(key);
-              break;
-            }
-          }
-        }
-      } else {
-        // this workspace should be displayed on all monitors
-        m_persistentWorkspacesToCreate.emplace_back(key);
-      }
-    }
-  }
-}
-
-void Workspaces::createPersistentWorkspaces() {
-  for (const std::string &workspaceName : m_persistentWorkspacesToCreate) {
-    Json::Value newWorkspace;
-    try {
-      // numbered persistent workspaces get the name as ID
-      newWorkspace["id"] = workspaceName == "special" ? -99 : std::stoi(workspaceName);
-    } catch (const std::exception &e) {
-      // named persistent workspaces start with ID=0
-      newWorkspace["id"] = 0;
-    }
-    newWorkspace["name"] = workspaceName;
-    newWorkspace["monitor"] = m_bar.output->name;
-    newWorkspace["windows"] = 0;
-    newWorkspace["persistent"] = true;
-
-    createWorkspace(newWorkspace);
-  }
-}
+void Workspaces::initializeWorkspaces() {
+  spdlog::debug("Initializing workspaces");
 
-void Workspaces::extendOrphans(int workspaceId, Json::Value const &clientsJson) {
-  for (const auto &client : clientsJson) {
-    if (client["workspace"]["id"].asInt() == workspaceId) {
-      registerOrphanWindow({client});
-    }
-  }
-}
-
-void Workspaces::init() {
-  m_activeWorkspaceName = (gIPC->getSocket1JsonReply("activeworkspace"))["name"].asString();
-
-  // get monitor ID from name (used by persistent workspaces)
-  m_monitorId = 0;
-  auto monitors = gIPC->getSocket1JsonReply("monitors");
-  auto currentMonitor = std::find_if(
-      monitors.begin(), monitors.end(),
-      [this](const Json::Value &m) { return m["name"].asString() == m_bar.output->name; });
-  if (currentMonitor == monitors.end()) {
-    spdlog::error("Monitor '{}' does not have an ID? Using 0", m_bar.output->name);
-  } else {
-    m_monitorId = (*currentMonitor)["id"].asInt();
+  // if the workspace rules changed since last initialization, make sure we reset everything:
+  for (auto &workspace : m_workspaces) {
+    m_workspacesToRemove.push_back(workspace->name());
   }
 
-  const Json::Value workspacesJson = gIPC->getSocket1JsonReply("workspaces");
-  const Json::Value clientsJson = gIPC->getSocket1JsonReply("clients");
+  // get all current workspaces
+  auto const workspacesJson = gIPC->getSocket1JsonReply("workspaces");
+  auto const clientsJson = gIPC->getSocket1JsonReply("clients");
 
   for (Json::Value workspaceJson : workspacesJson) {
     std::string workspaceName = workspaceJson["name"].asString();
     if ((allOutputs() || m_bar.output->name == workspaceJson["monitor"].asString()) &&
         (!workspaceName.starts_with("special") || showSpecial()) &&
         !isWorkspaceIgnored(workspaceName)) {
-      createWorkspace(workspaceJson, clientsJson);
+      m_workspacesToCreate.emplace_back(workspaceJson, clientsJson);
+    }
```

Comment:
And this is the other occurrence I mentioned, which [currently features the call to `extendOrphans` in an `else`](https://github.com/Syndelis/Waybar/blob/5bd831db191093d56d877872a081e222a22d9caa/src/modules/hyprland/workspaces.cpp#L672-L674)